woocommerce/sensei-certificates
# Add certificate data field hooks

Continuation #94
Fixes #93
Relates to #285, #284, #134, #116, #103, #102

### Changes proposed in this Pull Request

These changes add the following two filters:
- `sensei_certificate_data_fields` - This filter allows other plugins to add (or remove) data fields that appear in the Certificate Template editor.
- `sensei_certificate_data_field_value` - This filter is applied to the content of each of the data fields before they are written to the PDF. It provides the following five arguments:
  - `$field_value` - The field content (which will be written to the PDF).
  - `$field_key` - The name of the field the content is being written for. (This could be used to, for example, hide specific data fields under certain conditions.)
  - `$student` - a WP_User object, the student the certificate is being generated for.
  - `$course` - a WP_Post object, the course the certificate is being generated for. Can be null if used for preview.

### Testing instructions

* Add a new data field:

``` php
add_filter( 'sensei_certificate_data_fields', 'my_add_credit_hours_data_field' );
function my_add_credit_hours_data_field( $data_fields ) {
	$data_fields['credit_hours'] = array(
		'type'                 => 'text',
		'name'                 => 'Credit Hours',
		'position_label'       => 'Credit Hours Position',
		'position_description' => 'Optional position of the Credit Hours text',
		'text_label'           => 'Credit Hours Text',
		'text_description'     => 'Text to display in the Credit Hours area.',
		'text_placeholder'     => 'Credit Hours: {{credit_hours}}',
	);

	return $data_fields;
}
```

* Enable the `{{credit_hours}}` template tag:

``` php
add_filter( 'sensei_certificate_data_field_value', 'my_replace_credit_hours_template_tag', 10, 5 );
function my_replace_credit_hours_template_tag( $field_value, $field_key, $student, $course = null ) {
	$template_tag = '{{credit_hours}}';

	if ( false === strpos( $field_value, $template_tag ) ) {
		return $field_value;
	}

	$credit_hours = $course
		? (int) get_post_meta( $course->ID, '_credit_hours', true )
		: 0; // Default value used for preview.

	return str_replace( $template_tag, $credit_hours, $field_value );
}
```

* Make sure the new field is present in the certificate template.
* Position the new field in the template and set the font, color, etc.
* Add a `_credit_hours` post meta with a numeric value to a course.
* Assign the template to the course.
* Complete the course.
* Open the course certificate.
* Make sure the credit hours meta value is present.
* Try changing the fonts and make sure they work.
* Make sure all other fields work as expected.

![image](https://github.com/user-attachments/assets/bdbdb762-25ee-4911-8b22-1d9885d79140)
![image](https://github.com/user-attachments/assets/e6e42d93-3f1b-452b-9cb4-faa6594d407b)
![image](https://github.com/user-attachments/assets/fc837401-1b3b-4438-9c7c-d126ad6a4889)


### New Hooks

* `sensei_certificate_data_fields` - This filter allows adding (or removing) data fields that appear in the Certificate Template editor.
* `sensei_certificate_data_field_value` - This filter is applied to the content of each of the data fields before they are written to the PDF.
